src/hooks/utility/explain.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2023-2024 Retake, Inc.
+//
+// This file is part of ParadeDB - Postgres for Search and Analytics
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::ffi::{CStr, CString};
+use std::time::Instant;
+
+use anyhow::Result;
+use pgrx::{error, pg_sys};
+
+use super::parse_query_from_utility_stmt;
+use crate::{
+    duckdb::connection,
+    hooks::query::{get_query_relations, is_duckdb_query, set_search_path_by_pg},
+};
+
+enum Style {
+    Postgres,
+    Duckdb,
+}
+struct ExplainState {
+    analyze: bool,
+    style: Style,
+}
+
+pub fn explain_query(
+    query_string: &core::ffi::CStr,
+    stmt: *mut pg_sys::ExplainStmt,
+    dest: *mut pg_sys::DestReceiver,
+) -> Result<bool> {
+    let query = unsafe { (*stmt).query as *mut pg_sys::Query };
+
+    let query_relations = get_query_relations(unsafe { (*query).rtable });
+    if unsafe { (*query).commandType } != pg_sys::CmdType::CMD_SELECT
+        || !is_duckdb_query(&query_relations)
+    {
+        return Ok(true);
+    }
+
+    let state = parse_explain_options(unsafe { (*stmt).options });
+    let query = parse_query_from_utility_stmt(query_string)?;
+
+    let output = match state.style {
+        Style::Postgres => {
+            let mut output = format!("DuckDB Scan: {}\n", query);
+            if state.analyze {
+                let start_time = Instant::now();
+                set_search_path_by_pg()?;
+                connection::execute(&query, [])?;
+                let duration = start_time.elapsed();
+                output += &format!(
+                    "Execution Time: {:.3} ms\n",
+                    duration.as_micros() as f64 / 1_000.0
+                );
+            }
+            output
+        }
+        Style::Duckdb => {
```

Comment:
This is cool. What do you see as the use case for supporting both styles? Is there anything the DuckDB style offers that can't be in a single style?

I don't want to increase cognitive load on the user. Ideally, they wouldn't need to pass any setting to EXPLAIN if they can avoid it. What do you think?

Reply:
<img width="727" alt="image" src="https://github.com/user-attachments/assets/4f033a40-9ecd-4d88-939d-577df9c0dcad">
<img width="730" alt="image" src="https://github.com/user-attachments/assets/483a15cf-f154-40dd-9b60-e8b9541aca1b">

Reply:
That looks super clean, I really like that. Great work, wow! We should document this in the `docs/` in paradedb/paradedb.

Really nice stuff. I assume it works with (analyze) even without the ()? i.e. `EXPLAIN ANALYZE`?